source/common/http/conn_manager_impl.h

```diff
@@ -204,26 +204,16 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     std::list<AccessLog::InstanceSharedPtr> accessLogHandlers() override {
       return filter_manager_.accessLogHandlers();
     }
-    // Copy this stream's StreamInfo into the deferred logging object, ensuring that the request
+    // Copy this stream's StreamInfo into the response encoder, ensuring that the request
     // headers pointed to by the StreamInfo lives beyond stream destruction.
     void deferHeadersAndTrailers() {
-      Http::DeferredLoggingHeadersAndTrailers headers_and_trailers;
-      if (requestHeaders()) {
-        headers_and_trailers.request_header_map = request_headers_;
-      }
-      if (responseHeaders()) {
-        headers_and_trailers.response_header_map = response_headers_;
-      }
-      if (responseTrailers()) {
-        headers_and_trailers.response_trailer_map = response_trailers_;
-      }
       std::unique_ptr<StreamInfo::StreamInfoImpl> stream_info =
           std::make_unique<StreamInfo::StreamInfoImpl>(
               connection_manager_.codec_->protocol(), connection_manager_.time_source_,
               connection_manager_.read_callbacks_->connection().connectionInfoProviderSharedPtr());
       stream_info->setFrom(streamInfo(), request_headers_.get());
-      headers_and_trailers.stream_info = std::move(stream_info);
-      response_encoder_->setDeferredLoggingHeadersAndTrailers(headers_and_trailers);
+      response_encoder_->setDeferredLoggingHeadersAndTrailers(
+          request_headers_, response_headers_, response_trailers_, std::move(stream_info));
```

Comment:
nice! One more nit: can you pass in a const reference of the current stream info and create the new one in setDeferredLoggingHeadersAndTrailers()?